File: src/main/java/com/braydenoneal/blang/parser/statement/StatementType.java
package com.braydenoneal.blang.parser.statement;

import com.mojang.serialization.Lifecycle;
import com.mojang.serialization.MapCodec;
import net.minecraft.registry.Registry;
import net.minecraft.registry.RegistryKey;
import net.minecraft.registry.SimpleRegistry;
import net.minecraft.util.Identifier;

public record StatementType<T extends Statement>(MapCodec<T> codec) {
    public static final Registry<StatementType<?>> REGISTRY = new SimpleRegistry<>(
            RegistryKey.ofRegistry(Identifier.of("blogic", "statement_types")), Lifecycle.stable()
    );
}

File: src/main/kotlin/com/braydenoneal/blang/parser/statement/StatementType.kt
package com.braydenoneal.blang.parser.statement

import com.mojang.serialization.Lifecycle
import com.mojang.serialization.MapCodec
import net.minecraft.registry.Registry
import net.minecraft.registry.RegistryKey
import net.minecraft.registry.SimpleRegistry
import net.minecraft.util.Identifier

data class StatementType<T : Statement>(val codec: MapCodec<T>) {
    companion object {
        val REGISTRY: Registry<StatementType<*>> = SimpleRegistry(
            RegistryKey.ofRegistry(Identifier.of("blogic", "statement_types")), Lifecycle.stable()
        )
    }
}
